torch/csrc/jit/fusion_compiler.cpp

```diff
@@ -1047,67 +1140,409 @@ struct CPUFusionFunction : public CompiledFusionFunction {
+    if (producer->kind() == prim::Param) {
+      inputs.insert(val);
+      continue;
+    }
+    for (Value * input : producer->inputs()) {
+      if (/*bool inserted = */seen.insert(input).second) {
+        queue.push_back(input);
+      }
+    }
+  }
+
+  // Convert Value* into offsets into the graph's input list
+  std::vector<int64_t> offsets;
+  offsets.reserve(inputs.size());
+  for (Value * input : inputs) {
+    offsets.push_back(input->offset());
+  }
+  std::sort(offsets.begin(), offsets.end());
+  return offsets;
+}
+
+std::vector<std::vector<int64_t>> FusedKernelCache::getInputBroadcastGroups() {
+  std::unordered_set<std::vector<int64_t>, torch::hash<std::vector<int64_t>>> broadcast_groups;
+  for (Value * output : graph->outputs()) {
+    broadcast_groups.insert(getInputDependencies(output));
+  }
+  return std::vector<std::vector<int64_t>>{ broadcast_groups.begin(), broadcast_groups.end() };
+}
+
+void FusedKernelCache::run(Stack& stack) {
+  int64_t num_inputs = graph->inputs().size();
+  auto args = fmap(last(stack, num_inputs), [](const IValue& i) {
+                return i.toTensor();
+              });
+
+  auto maybe_map_size = canRunKernel(args);
+  if (!maybe_map_size) {
+    return runFallback(stack);
+  }
+  expandArgs(args, *maybe_map_size);
+
+  FusedKernelArgSpec spec { args };
+  auto it = kernels.find(spec);
+  if (it == kernels.end()) {
+    std::tie(it, std::ignore) = kernels.emplace(spec, compileSpec(spec, *maybe_map_size));
+  }
+  auto & fn = it->second;
+
+  std::vector<at::Tensor> outputs;
+  fn->launch(args, outputs);
+  drop(stack, num_inputs);
+  pack(stack, std::move(outputs));
+}
+
+at::optional<std::vector<int64_t>> FusedKernelCache::getMapSize(at::TensorList args, at::IntList arg_subset) {
+  // NB: we leave this uninitialized, because an empty size is trivially
+  // broadcastable to any other size.
+  // TODO: this keeps reallocating map_size at every iteration, but we know
+  // exactly how much storage do we need, so this could be fixed in-place at
+  // every step. We're just missing a few functions for ATen, but the fix
+  // should be straightforward.
+  int64_t dim_after_broadcast = 0;
+  for (int64_t arg_idx : arg_subset) {
+    dim_after_broadcast = std::max(dim_after_broadcast, args[arg_idx].dim());
+  }
+  std::vector<int64_t> map_size;
+  for (size_t i = 0; i < arg_subset.size(); ++i) {
+    auto & arg = args.at(arg_subset[i]);
+    auto & chunk_desc = input_chunks.at(arg_subset[i]);
+    if (chunk_desc.nSubtensors == 1) {
+      try {
+        map_size = at::infer_size(map_size, arg.sizes());
+      } catch (std::exception& e) {
+        return at::nullopt;
+      }
     } else {
-      JIT_ASSERT(canCompileOnCPU());
-      raw_func = new CPUFusionFunction(name, agraph, config_);
+      auto tensor_sizes = arg.sizes().vec();
+      int64_t num_chunks = chunk_desc.nSubtensors;
+      int64_t dim = chunk_desc.dim;
+      if (dim < 0) {
+        dim += arg.dim();
```

Comment:
Shouldn't you just use the wrap dims utils here?

Reply:
Sure, will use that.